src/lazy/any_encoding.rs

```diff
@@ -0,0 +1,535 @@
+        LazyRawAnySequence {
+            encoding: LazyRawSequenceKind::Text_1_0(value),
+        }
+    }
+}
+
+impl<'data> From<LazyRawBinarySequence<'data>> for LazyRawAnySequence<'data> {
+    fn from(value: LazyRawBinarySequence<'data>) -> Self {
+        LazyRawAnySequence {
+            encoding: LazyRawSequenceKind::Binary_1_0(value),
+        }
+    }
+}
+
+// ===== Structs =====
+
+#[derive(Debug, Clone)]
+pub struct LazyRawAnyStruct<'data> {
+    encoding: LazyRawStructKind<'data>,
+}
+
+#[derive(Debug, Clone)]
+pub enum LazyRawStructKind<'data> {
+    Text_1_0(LazyRawTextStruct<'data>),
+    Binary_1_0(LazyRawBinaryStruct<'data>),
+}
+
+#[derive(Debug, Clone)]
+pub struct LazyRawAnyField<'data> {
+    encoding: LazyRawFieldKind<'data>,
+}
+
+#[derive(Debug, Clone)]
+pub enum LazyRawFieldKind<'data> {
+    Text_1_0(LazyRawTextField<'data>),
+    Binary_1_0(LazyRawBinaryField<'data>),
+}
+
+pub struct RawAnyStructIterator<'data> {
+    encoding: RawAnyStructIteratorKind<'data>,
+}
+
+pub enum RawAnyStructIteratorKind<'data> {
+    Text_1_0(RawTextStructIterator<'data>),
+    Binary_1_0(RawBinaryStructIterator<'data>),
+}
+
+impl<'data> Iterator for RawAnyStructIterator<'data> {
+    type Item = IonResult<LazyRawAnyField<'data>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match &mut self.encoding {
+            RawAnyStructIteratorKind::Text_1_0(i) => i
+                .next()
+                .map(|field_result| field_result.map(|field| field.into())),
+            RawAnyStructIteratorKind::Binary_1_0(i) => i
+                .next()
+                .map(|field_result| field_result.map(|field| field.into())),
+        }
+    }
+}
+
+impl<'data> From<LazyRawTextField<'data>> for LazyRawAnyField<'data> {
+    fn from(value: LazyRawTextField<'data>) -> Self {
+        LazyRawAnyField {
+            encoding: LazyRawFieldKind::Text_1_0(value),
+        }
+    }
+}
+
+impl<'data> From<LazyRawBinaryField<'data>> for LazyRawAnyField<'data> {
+    fn from(value: LazyRawBinaryField<'data>) -> Self {
+        LazyRawAnyField {
+            encoding: LazyRawFieldKind::Binary_1_0(value),
+        }
+    }
+}
+
+impl<'data> LazyRawFieldPrivate<'data, AnyEncoding> for LazyRawAnyField<'data> {
+    fn into_value(self) -> LazyRawAnyValue<'data> {
+        match self.encoding {
+            LazyRawFieldKind::Text_1_0(f) => f.value.into(),
+            LazyRawFieldKind::Binary_1_0(f) => f.value.into(),
+        }
+    }
+}
+
+impl<'data> LazyRawField<'data, AnyEncoding> for LazyRawAnyField<'data> {
+    fn name(&self) -> RawSymbolTokenRef<'data> {
+        match &self.encoding {
+            LazyRawFieldKind::Text_1_0(f) => f.name(),
+            LazyRawFieldKind::Binary_1_0(f) => f.name(),
+        }
+    }
+
+    fn value(&self) -> LazyRawAnyValue<'data> {
+        match &self.encoding {
+            LazyRawFieldKind::Text_1_0(f) => f.value().into(),
+            LazyRawFieldKind::Binary_1_0(f) => f.value().into(),
+        }
+    }
```

Comment:
🗺️ There's a subtle API challenge here.

Prior to this PR, the `value()` trait method required that implementers return a `&D::Value`; in the case of `AnyEncoding`, that would mean returning a `&LazyRawAnyValue`.

Any time a method returns a `&T` (not `&'static T`), there MUST be a `T` living inside one of the method's parameters. In this case, there would need to be a `LazyRawAnyValue` living inside `self`, which there isn't. Instead, `self` contains the raw materials to construct a `LazyRawAnyValue`: either a `LazyRawTextValue` or a `LazyRawBinaryValue`. We can't return a `&LazyRawAnyValue` because it's illegal to return a reference to a stack-allocated variable created in this method's scope.

Fortunately, both `LazyRawTextValue` and `LazyRawBinaryValue` are eligible to implement `Copy`. Neither owns any heap-allocated resources. I changed the trait to return a copy of the value instead of a reference to one, resolving this issue. There are minor ripples of this change through the rest of this PR.

Reply:
Now that `as_value` is returning a copy of the value and not a reference, it should be called `to_value`. I've captured that in #626 and will address it when the merge queue is empty.